### PR TITLE
Inventory sub dialog graph

### PR DESCRIPTION
**Pull request by N-thony**
@rdstern I made improvements to the inventory graph. Can you confirm that the points you raised about issue #6029 have been made? Thank you.

### CODE REVIEW

**Comment by rdstern** [comment]
@N-thony this is a good start.

1. On the sub-dialogue you have, I note that the x-axis "Year" is there initially.  Then it is absent after using the sub-dialogue.
2. On the initial titles sub-dialogue the font-sizes are all zero.  If you look at the Climatic > PICSA  > Rainfall graph sub-dialogue, they start at 20, 15, 8.
3. Again, on the PICSA sub-dialogue, there is a useful tooltip in the footnote about `\n` giving a newline.  That would be good to add.
4. What you have in the y-axis and x-axis tabs could be added, instead, into the Titles tab.  Again, as it is in the PICSA > Rainfall sub-dialogue.
5. Then the x-axis tab is mainly for the spacing of the years as they are labelled there.  See below.
image

For example if I would like those labels to be every 5 years, instead of every 20 years, I would do it there.
6. The y-axis may be more complicated and you will need help and advice from @shadrackkibet and @dannyparsons but the example above shows that it isn't great currently.  In this example, I can't read the information on either the right or the left of the y-axis.

Thanks.

Roger

**Comment by N-thony** [comment]
@rdstern I solved the issues you raised days ago. Could you please confirm the issues are resolved? Thanks.

**Comment by rdstern** [comment]
Thanks.  Will do.

File: instat/static/InstatObject/R/data_object_R6.R [hunk 1]
```diff
@@ -2608,11 +2616,14 @@ DataSheet$set("public","make_inventory_plot", function(date_col, station_col = N
...
     }
-  }
+   if(scale_xdate){ g <- g + ggplot2::scale_x_date(date_minor_breaks = paste0(byXaxis," year"), limits = c(from=as.Date(paste0(fromXAxis,"-01-01")), to = as.Date(paste0(toXAxis,"-12-31")))) } 
+ }
   if(coord_flip) {
     g <- g + ggplot2::coord_flip()
   }
-  return(g + ggplot2::ggtitle(graph_title) + ggplot2::theme(plot.title = ggplot2::element_text(hjust = 0.5)))
+  if(!is.null(labelXAxis)){g <- g + ggplot2::xlab(labelXAxis)}else{g <- g + ggplot2::xlab("")}
+  if(!is.null(labelYAxis)){g <- g + ggplot2::ylab(labelYAxis)}else{g <- g + ggplot2::ylab("")}
```
Comment (shadrackkibet) [inline instat/static/InstatObject/R/data_object_R6.R:2625]: Always use `missing()` to check if an argument is specified. `is.null`  is a check for the reserved word "NULL" which means "empty". Then use ` ggplot2::xlab(NULL)` and `ggplot2::ylab(NULL)` which gives more space on the axis.

**Review by shadrackkibet — CHANGES_REQUESTED** [review]

File: instat/static/InstatObject/R/data_object_R6.R (hunk 1 above)
Comment (N-thony) [reply instat/static/InstatObject/R/data_object_R6.R]: Thanks

**Comment by N-thony** [comment]
@shadrackkibet I solved the issue you raised days ago. Could you please confirm the issue is resolved? Thanks.

File: instat/static/InstatObject/R/data_object_R6.R [hunk 2]
```diff
@@ -2575,23 +2579,27 @@ DataSheet$set("public","make_inventory_plot", function(date_col, station_col = N
       }
     }
     else if(length(element_cols) > 1) {
-      g <- g + ggplot2::facet_grid(facets = variable~.)
+          if(scale != "none"){
+               g <- g + ggplot2::facet_wrap(.~variable, nrow = nrow, ncol = ncol, scales = scale, dir = dir)
+	      }else if(scale == "none"){g <- g + ggplot2::facet_grid(facets = variable~.)}
     }
+     if(scale_xdate){ g <- g + ggplot2::scale_x_continuous(breaks=seq(from = fromXAxis, to = toXAxis, by = byXaxis)) } 
+     if(scale_ydate && !is.null(date_ybreaks) && !is.null(date_ylabels)){ g <- g + ggplot2::scale_y_date(date_breaks = paste0(step," ",date_ybreaks), date_labels = date_ylabels) }
```
Comment (shadrackkibet) [inline instat/static/InstatObject/R/data_object_R6.R:2587]: What about this? Check other places as well.

**Review by shadrackkibet — CHANGES_REQUESTED** [review]

**Comment by N-thony** [comment]
@shadrackkibet I solved the issue you raised minutes ago. Could you please confirm the issue is resolved? Thanks.

**Comment by lloyddewit** [comment]
@rdstern are the issues you raised above now resolved? thanks

**Comment by rdstern** [comment]
@N-thony I tried to check the sub-dialogues, but they are still disabled.  I can't check until they are enabled.  Thanks

**Comment by N-thony** [comment]
> @N-thony I tried to check the sub-dialogues, but they are still disabled. I can't check until they are enabled. Thanks

@rdstern what do you mean by they are still disabled? The Sub-dialog or something else?

**Comment by rdstern** [comment]
@N-thony sorry, I can get the Inventory sub-dialogue.

I still can't sort out the x-axis or y-axis scales as I would like to.  On x-axis it is usually like the PICSA sub-dialogue.  There I can do this

image

So, I am not sure you need None and Date.  It is usually (I haven't tried all options, just numeric, so the same as PICSA should work.  See the year x-axis in the picture above.

You have allowed for dates so I tried them.  I am not sure it is needed for this, but might be for the y-axis.  When I tried with date, because that was my only option to get more labels so:
image

Gave:
image

As I say - I am not sure you need the date option for the x.

**Comment by rdstern** [comment]
Here is an example for a possible date change on the y-axis
image
Possibly that might not give much flexibility, but could allow anything from one month to 6 months, and also allow the font to change.  Also just possibly have none?  What do you think?

File: instat/instat.vbproj [hunk 3]
```diff
@@ -2741,6 +2747,9 @@
     <EmbeddedResource Include="dlgTreemap.resx">
       <DependentUpon>dlgTreemap.vb</DependentUpon>
     </EmbeddedResource>
+    <EmbeddedResource Include="sdgInventoryGraph.resx">
+      <DependentUpon>sdgInventoryGraph.vb</DependentUpon>
+    </EmbeddedResource>
```
Comment (shadrackkibet) [inline instat/instat.vbproj:2752]: You don't have to create a new sub dialog. There is already an existing form (sub dialog) `sdgInventoryGraph.vb`. So I don't expect changes in `.vbproj `

File: instat/dlgInventoryPlot.vb [hunk 4]
```diff
@@ -327,7 +322,11 @@ Public Class dlgInventoryPlot
         End If
 
     End Sub
-
+    Private Sub cmdInventoryPlotOptions_Click(sender As Object, e As EventArgs) Handles cmdInventoryPlotOptions.Click
+        sdgInventoryPlot.SetRCode(clsInventoryNew:=clsInventoryPlot, bReset:=bResetSubdialog)
+        sdgInventoryPlot.ShowDialog()
```
Comment (shadrackkibet) [inline instat/dlgInventoryPlot.vb:327]: You don't have to create a new sub dialog. There is already an existing form (sub dialog) `sdgInventoryGraph.vb`. So I don't expect changes in `.vbproj`

**Comment by rdstern** [comment]
image

This is great, just what I was hoping for in flexibility.

image
or
image
or
image

a) In the y-axis tab the breaks up-down can go down to zero.  It gives an error if it is zero.  Make 1 the minimum.
b) In the Size, the up-down is inactive.  It is still usable, as you can type in (which I like), but the up-down should work.
 
c) A curiosity, that perhaps @dannyparsons can help to explain is the Jan appearing twice?  And the numbers in the y-axis are slightly confusing in starting (for example above) from 23 Dec and going to 17 Jan?
d) On the facets tab the first option is size.  Could there be two controls here for Station and Element.  In the example above I would like quite large, or default for the stations (only 2 stations), but quite small for the elements.  I assume this is possible?

To me a big improvement is the x-axis options and that is working well.  Great.

**Comment by N-thony** [comment]
@rdstern I solved the issues you raised last Friday. Could you please confirm that the issues are resolved? Thanks

**Comment by rdstern** [comment]
Looks good:
image

On the labelling I notice that the axis labels are (I think) "above" the date.  I wonder if they can be centred instead? (Or their position could become an option?  @N-thony how is your R or this or do you want to check with @shadrackkibet

**Comment by N-thony** [comment]
@rdstern I solved the issue you raised minutes ago. Could you please confirm that the issue is resolved? Thanks

**Review by rdstern — DISMISSED** [review]
I am happy to approve this now, because it is a considerable improvement by @N-thony to this important dialogue.  He has added considerable flexibility to the dialogue.  The original is available and this is an improvement.
I hope @lloyddewit could check the text, but would also like either @shadrackkibet or @lilyclements to check on a remaining problem and also look at the design of the sub-dialogue tabs to see if they can suggest further options, and/or label and layout improvements.

image

Here is the remaining problem (or problems).  When looking by year it seems to have 13 categories, ostensibly going from January to January.  I think the data may be ok and this is just ggplot being a bit clever on the labelling in a way we don't want here.  This is a general issue on this option.
2. A special problem is when the rain days are displayed.  We don't need to keep this option in the dialogue.  But it occasionally seems to give a problem.  Here it looks as though December 2000 was all rainy.  

Here are the data I was using:  
[ghana_2_stations.zip](https://github.com/africanmathsinitiative/R-Instat/files/5730463/ghana_2_stations.zip)

File: instat/sdgInventoryPlot.vb [hunk 5]
```diff
@@ -1,3 +1,328 @@
...
+'
+' You should have received a copy of the GNU General Public License 
+' along with this program.  If not, see <http://www.gnu.org/licenses/>.
 
+Imports instat
+Imports instat.Translations
+Public Class sdgInventoryPlot
+    Private bControlsInitialised As Boolean = False
+    Private clsInventory As New RFunction
+    Public strAxisType As String
```
Comment (lloyddewit) [inline instat/sdgInventoryPlot.vb:22]: Is this needed?

File: instat/sdgInventoryPlot.vb [hunk 6]
```diff
@@ -1,3 +1,328 @@
...
+    Private Sub ucrPnlHorizonatalVertical_ControlValueChanged(ucrChangedControl As ucrCore) Handles ucrPnlHorizonatalVertical.ControlValueChanged
+        SetFacetParameter()
+
+    End Sub
+
+    Private Sub ucrChkNoOfRowsorColumns_ControlValueChanged(ucrChangedControl As ucrCore) Handles ucrChkNoOfRowsorColumns.ControlValueChanged
+        SetFacetParameter()
+    End Sub
+
+    Private Sub UcrNud1_Load(sender As Object, e As EventArgs) Handles ucrNudFacetYSize.Load
```
Comment (lloyddewit) [inline instat/sdgInventoryPlot.vb:325]: Is this function needed?

**Review by lloyddewit — CHANGES_REQUESTED** [review]
@shadrackkibet Thank you for your peer review
@N-thony This is good work! I just have a few small suggestions

**Comment by N-thony** [comment]
@lloyddewit I solved the issues you raised hours ago. Could you confirm that the issues are resolved? Thanks.

File: instat/static/InstatObject/R/data_object_R6.R [hunk 7]
```diff
@@ -2455,7 +2455,7 @@ DataSheet$set("public","append_climatic_types", function(types) {
...
+DataSheet$set("public","make_inventory_plot", function(date_col, station_col = NULL, year_col = NULL, doy_col = NULL, element_cols = NULL, add_to_data = FALSE, year_doy_plot = FALSE, coord_flip = FALSE, facet_by = NULL, facet_xsize = 7, facet_ysize = 11, graph_title = "Inventory Plot", graph_subtitle = NULL, graph_caption = NULL, title_size = NULL, subtitle_size = NULL, caption_size = NULL, labelXAxis, labelYAxis, xSize = NULL, ySize = NULL, Xangle = NULL, Yangle = NULL, fromXAxis = NULL, toXAxis = NULL, byXaxis = NULL, date_ylabels, scale_xdate = FALSE, legend_position = NULL, xlabelsize = NULL, ylabelsize = NULL, scale = "none", dir = NULL, nrow = NULL, ncol = NULL, scale_ydate = FALSE, date_ybreaks, step = 1, key_colours = c("red", "grey"), display_rain_days = FALSE, rain_cats = list(breaks = c(0, 0.85, Inf), labels = c("Dry", "Rain"), key_colours = c("tan3", "blue"))) {
```
[the remark this answers is not in the document]
Comment (lloyddewit) [reply instat/static/InstatObject/R/data_object_R6.R]: Thanks for updating this line. There are other examples of very long lines in this function but they are throughout the R code. So it's OK to leave them as is for now.

**Comment by lloyddewit** [comment]
@N-thony thanks, all my comments above are resolved.
I am ready to approve but there is still an issue with `instat.vbproj`. I resolved all the conflicts and the file in this PR is now identical to the file on the `master` branch (I double-checked this with an external diff tool). However, GitHub is confused and thinks that every line has changed. I cannot correct this myself because I don't seem to have push privileges to this PR.
It would be best if `instat.vbproj` disapeared totally from this PR. Please could you try one more time to copy the file from `master` and push it to this PR? There are a couple of ways of doing this, we can discuss if it helps.

**Comment by lloyddewit** [comment]
@dannyparsons I am ready to approve this PR but there is an issue with 'instat.vbproj'. This was updated in this PR but the changes were not needed so we tried to reverse them. As far as I can see, the file in this PR is now identical to the file on the master branch (I checked this with an external diff tool). However, GitHub shows nearly every line as changed. 
We could go ahead and merge but it would be better if 'instat.vbproj' disapeared completely from this PR.
I could research how to achieve this but it may be time-consuming. 
Have you come across this type of situation before? Do you have any advice?
thanks

**Comment by lloyddewit** [comment]
@N-thony I think the easiest solution to the `instat.vbproj` issue is to close this PR and make a new one.
Please could you:
- Make a new branch from the current master branch
- Copy the changes from this PR to the new branch (all files except `instat.vbproj`)
- Submit a new PR (which references this PR)
- Close this PR
Please let me know if you have any questions, thanks

**Pull request closed without merging**